libs/vkd3d/swapchain.c

```diff
@@ -2409,6 +2613,178 @@ static CONST_VTBL struct IDXGIVkSwapChainFactoryVtbl dxgi_vk_swap_chain_factory_
+        latency_sleep_info.pNext = NULL;
+        latency_sleep_info.signalSemaphore = chain->present.low_latency_sem;
+        latency_sleep_info.value = chain->present.low_latency_sem_value;
+
+        pthread_mutex_lock(&chain->present.low_latency_lock);
+
+        if (chain->present.vk_swapchain)
+        {
+            should_sleep = true;
+            VK_CALL(vkLatencySleepNV(chain->queue->device->vk_device, chain->present.vk_swapchain, &latency_sleep_info));
+        }
+
+        pthread_mutex_unlock(&chain->present.low_latency_lock);
+
+        if (should_sleep)
+        {
+            memset(&sem_wait_info, 0, sizeof(sem_wait_info));
+            sem_wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO;
+            sem_wait_info.pNext = NULL;
+            sem_wait_info.flags = 0;
+            sem_wait_info.semaphoreCount = 1;
+            sem_wait_info.pSemaphores = &chain->present.low_latency_sem;
+            sem_wait_info.pValues = &chain->present.low_latency_sem_value;
+
+            VK_CALL(vkWaitSemaphores(chain->queue->device->vk_device, &sem_wait_info, UINT64_MAX));
+        }
+    }
+}
+
+void dxgi_vk_swap_chain_set_latency_sleep_mode(struct dxgi_vk_swap_chain* chain, bool low_latency_mode,
+	bool low_latency_boost, uint32_t minimum_interval_us)
+{
+    /* If the low latency state is already set to what the application is
+     * requesting, and there isn't an update in flight it is safe to skip
+     * this request. */
+    if (chain->present.low_latency_update_requested == false &&
+            low_latency_mode == chain->present.low_latency_state.mode &&
+            low_latency_boost == chain->present.low_latency_state.boost &&
+            minimum_interval_us == chain->present.low_latency_state.minimum_interval_us)
+        return;
+
+    pthread_mutex_lock(&chain->present.low_latency_lock);
+
+    chain->present.requested_low_latency_state.mode = low_latency_mode;
+    chain->present.requested_low_latency_state.boost = low_latency_boost;
+    chain->present.requested_low_latency_state.minimum_interval_us = minimum_interval_us;
+
+    /* The actual call to vkSetLatencySleepModeNV will happen
+     * in the present task when the swapchain is recreated. This
+     * is required when enabling low latency mode to use the frameId
+     * provided by the application, because the id it provides may be
+     * less than the largest id already used with the current swapchain. */
+    chain->present.low_latency_update_requested = true;
+    chain->present.force_swapchain_recreation = true;
+
+    pthread_mutex_unlock(&chain->present.low_latency_lock);
+}
+
+void dxgi_vk_swap_chain_set_latency_marker(struct dxgi_vk_swap_chain* chain, uint64_t frameID, VkLatencyMarkerNV marker)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &chain->queue->device->vk_procs;
+    VkSetLatencyMarkerInfoNV latency_marker_info;
+
+    if (chain->present.low_latency_state.mode && chain->present.using_low_latency_present_mode)
```

Comment:
Racy to check this here.

Reply:
We don't really need this check so I just ended up removing it. It is fine to provide the driver with latency markers when low latency mode is disabled.